### ISSUE TITLE
struggling to get accurate enter/leave events from ibeacon

### ISSUE THREAD

**Issue by josh-m-sharpe** [owntracks/ios#724]
Use case:
My ibeacon is pluged into a 12v -> USB adapter in my car.  The goal is to let Home Assistant know if I'm in my car.  I don't need to know where I am, just that my phone is physically near the ibeacon and then push some version of true/false to mqtt -> Home Assistant.

Here's my current config:

![Notification_Center](https://user-images.githubusercontent.com/39473/198147785-72025881-3ace-4f18-ae88-dab13a1dbcd6.jpg)

Some problems I've seen recently:
1) Left my office out to my car and started it up.  Home Assistant didn't show I was in my car.  I opened the `iBeacon` app and started scanning.  The beacon was detected easily.  I then opened Owntracks and saw that Owntracks hadn't turned the icon from Blue to Red indicating it wasn't detected.  After flipping to the map view and back, it changed red, and then Home Assistnat showed I was in my car.  Yay!
2) Drove my car to my office.  Got out, turned the car off, walked 100+ft to the door, then up a 5 story elevator.  Home Assistant showed I was still in my car.  I opened Owntracks and saw that the beacon was still red on the Regions and Map screen.  I then opened iBeacon and scanned -- nothing detected (obviously out of range).  But OwnTracks didn't pick that up?
3) Drove all the way home, parked my car, went inside.  Opened Owntracks to see that the beacon was Red.  Opened iBeacon and scanned which didn't find anything (this is expected inside the house) -- but why did owntracks still show that it was "red".

So, there's a variety of conditions in which owntracks appears to not be properly reflecting whether I'm near, or not near, the beacon.  However, If I go in and 'fiddle' with it, then eventually it'll update, and send the MQTT event to Home Assistant.

I'm curious why this isn't happening automagically in the background.  Is this a limitation of iOS killing the app and being lame?

Thanks!

**Comment by ckrey** [comment]
Please check if the OwnTracks app is allowed to always use location (Settings/Privacy/Location Services). Detecting iBeacons in Background depends on that setting.

It is not the app detecting entering/exiting an iBeacon, but iOS, which informs the app.
The detection usually works reliably within a few seconds.

**Comment by josh-m-sharpe** [comment]
Location enabled:
![IMG_13FCAE434F7A-1](https://user-images.githubusercontent.com/39473/198267262-d7dffe3d-3e3b-4a93-b906-f9c0177de0ba.jpeg)

Significant Mode:

![IMG_6DAF8AB54CC2-1](https://user-images.githubusercontent.com/39473/198267625-1f0c6ce4-49fc-4b80-bba0-b1625cdbaae5.jpeg)

**Comment by josh-m-sharpe** [comment]
> Please check if the OwnTracks app is allowed to always use location (Settings/Privacy/Location Services). Detecting iBeacons in Background depends on that setting.
> 
> Please check on the Map screen if you are in Manual, Significant, or Move mode. In Quiet mode all location or region detection is turned off.
> 
> It is not the app detecting entering/exiting an iBeacon, but iOS, which informs the app. The detection usually works reliably within a few seconds.

Understood that it's iOS doing the detecting.   I guess what would explain the scenario of me taking 2-3 minutes to walk in my office building and go up a 5 story elevator and *still* see it Red in OwnTracks (but not scannable from iBeacon)?

**Comment by josh-m-sharpe** [comment]
fwiw, other test I'm doing:

Sitting in my house with the iBeacon next to me.  
All the apps are open on my phone - but iOS is on the home screen and the phone is locked... so all apps backgrounded.
Monitoring Home Assistant on my laptop for the sensor I have set up to change.

If I unplug or plug in the beacon it works great.  It's not instantaneous, but like you said a few seconds and it changes.  Works both ways.

So, it works in ideal situations -- something's just getting out of wack out in the wild.

**Comment by josh-m-sharpe** [comment]
Is there anything I can do when I notice it hasn't been detected while in the background?

My concern is that opening the app changes things, or triggers it to re-check things... I dunno.  Basically feels like a catch-22 opening the app since normally I'd never do that.

Thanks again!

**Comment by ckrey** [comment]
Hi Josh, I did install an iBeacon in my car too. Works perfectly and sends out MQTT transition events and locations immediately.

When you say "while in background", do you mean iPhone is locked or swiped out the app?

Would you share your exported settings (without the credentials) with out via support@owntracks.org ?

**Comment by josh-m-sharpe** [comment]
Under normal circumstances I would never open Owntracks or kill it.  So, "while in background" means I opened it, configured it, then switched out of it.  It sits alive in the background indefinitely.

Since my last comment, I have updated my HomeAssistant configuration to send me a push notification every time OwnTracks sends a message ( this is a lot of messages! :) )

Some more observations:

1. The other morning I walked out of my house to my car, turned the car on, got a push notification that I was NOT in my car.  The first thing I did was open the `KBeacon` app and pressed `scan`.  I wanted to see if the beacon was alive and talking.  The KBeacon app immediately detected the beacon AND then almost immediately I got a HomeAssistant Notification that I was in my car. (which implies Owntracks in the background did things).
2.  As I sit here in my office Home Assistant thinks i'm in my car.  If I open Owntracks the beacon is Red implying that it's nearby but KBeacon doesn't detect the beacon.  I'm most certainly out of range.  In my office, from time to time, I get this notification.  It currently says "true"
3. As I drive down the highway, I frequently get these notifications.  Though, due to this problem sometime it is "true" and other times it is "false". 
4. I don't think I have ever opened `KBeacon` app and it incorrectly detected the beacon.  If I'm physically close to the beacon it shows up and otherwise it doesn't show up.

I wonder if walking into my office from my car isn't a "significant" enough change to wake OwnTracks up and send an alert?

However, even if that is true, the periodic updates throughout the day (currently that I'm in my car) suggest that when OwnTracks does in fact wake up, it doesn't check the status of the beacon's proximity and update accordingly?

I have mailed my settings to you.  Thanks for taking a look!

**Comment by josh-m-sharpe** [comment]
Update:  Owntracks must've woken up and reported that I'm not near the beacon. I've (sadly) barely moved out of a 100ft radius in my office building.  The beacon is 5 stores below me out in the parking lot.

Seems fairly random after reporting that I'm near it all day long.

**Comment by ckrey** [comment]
I checked your configuration. Please try to remove the leading dash `-` in the name. This may cause a delay not well suited for background detection.

**Comment by josh-m-sharpe** [comment]
```
If you use a - as the first character in your Waypoint description e.g. -MyBeacon, the app will ignore short disruptions of the connection to the beacon (hold down).
```

I've just not removed the `-` and will see what happens.  That said, based on that description of the feature, I really don't think that's what is happening here.


For quick context here, I have Owntracks/HomeAssistant/node-red all wired up to send me a Home
Assistant push notification every time an input boolean changes that is tied to detection of this beacon.  I find this beneficial for debugging since I don't have to open OwnTracks or the HomeAssistant apps (and potentially disturb some state) in order to see that the system has detected this change.


Today the beacon stayed in detected status for hours after leaving my car and going into the office. I was occasionally getting push notifications that I was in the car - signaling that OwnTracks had not registered I was no longer near the car.

A few days ago I was sitting in my car, the input boolean in HA was false.  I opened `KBeacon` and pressed scan.  I immediately got a push notification that I was in the car.

So, OwnTracks definitely can detect change, but it seems to hang onto a value long after it has changed.  It's almost as if when OwnTracks wakes up in the background it is either not properly detecting the presence or lack there-of of a beacon OR it isn't afforded enough time to do so by the OS.  That says, other times the beacon is detected and I receive a push immediately.


As stated above I've removed the `-` and will keep watching - but based on the description of that feature I'm not completely convinced that's the issue here. (Also, it seems like a useful feature to introduce *some* change delay.  I do not require *immediate* updates -- but I have seen instances where it flips back and forth while driving down the highway and presumably this feature would help mitigate that.)

Sorry for the long post and delay in responding.  Thank you for your help!

**Comment by josh-m-sharpe** [comment]
Sorry for being verbose -- just trying to narrow down the most obvious, reproducible issue here.  

Since my last post I was able to easily reproduce this:
```
A few days ago I was sitting in my car, the input boolean in HA was false. I opened KBeacon and pressed scan. I immediately got a push notification that I was in the car.
```

This morning, seemingly the opposite is happening:
Owntracks (and thus HA) think I'm in my car, but it's out in the driveway and I'm too far to detect the beacon with KBeacon.  Opening KBeacon doesn't detect anything, so my guess is iOS doesn't "wake the system up" and notify other apps since there's nothing to notify about.  I opened Owntracks and see that the region is Red, thus detected, and the app does nothing to change it back to blue - even foregrounded.  Now maybe just being in my house doesn't produce a significant location change ( i did get a push from HA though, which I think is only triggered by a significant location change).  But one would think that foregrounding the app would definitely set things to their current state?

Both scenarios seem to suggest that OwnTracks (sometimes) doesn't properly detect the beacon or absence of it when it wakes up and does its background thing.

Cheers and thank you for reading!
